src/effects/VST3/VST3EffectsModule.cpp

```
@@ -0,0 +1,320 @@
+#include "VST3EffectsModule.h"
+
+#include <stdexcept>
+
+#include <wx/stdpaths.h>
+#include <wx/dir.h>
+#include <wx/log.h>
+#include <public.sdk/source/vst/hosting/module.h>
+#include <public.sdk/source/vst/hosting/plugprovider.h>
+#include <pluginterfaces/vst/ivstaudioprocessor.h>
+
+
+#include "../../ModuleManager.h"
+
+#include "wxArrayStringEx.h"
+#include "PluginInterface.h"
+#include "ModuleInterface.h"
+#include "VST3Effect.h"
+#include "VST3Utils.h"
+#include "widgets/AButton.h"
+
+
+DECLARE_MODULE_ENTRY(AudacityModule)
+{
+   // Create our effects module and register
+   // Trust the module manager not to leak this
+   return safenew VST3EffectsModule();
+}
+
+DECLARE_BUILTIN_MODULE(VST3Builtin);
+
+//Helper class used to find vst3 effects during folder traversal
+class VST3PluginTraverser final : public wxDirTraverser
+{
+   std::function<void(const wxString&)> mOnPluginFound;
+public:
+   VST3PluginTraverser(std::function<void(const wxString&)> onPluginFound)
+      : mOnPluginFound(std::move(onPluginFound))
+   {
+   }
+
+   wxDirTraverseResult OnFile(const wxString& filename) override
+   {
+      if(filename.Matches("*.vst3"))
+         mOnPluginFound(filename);
+      return wxDIR_CONTINUE;
+   }
+   wxDirTraverseResult OnDir(const wxString& dirname) override
+   {
+      if(dirname.Matches("*.vst3"))
+      {
+         mOnPluginFound(dirname);
+         return wxDIR_IGNORE;//do not look inside...
+      }
+      return wxDIR_CONTINUE;
+   }
+};
+
+const TranslatableString VST3EffectsModule::PluginType = XO("VST3");
+
+std::shared_ptr<VST3::Hosting::Module> VST3EffectsModule::GetModule(const wxString& path)
+{
+   const auto it = mModules.find(path);
+   if(it != mModules.end())
+   {
+      if(auto lock = it->second.lock())
+         return lock;
+   }
+
+   std::string moduleCreateError;
+   auto module = VST3::Hosting::Module::create(path.ToStdString(), moduleCreateError);
+   if(!module)
+      throw std::logic_error(moduleCreateError.c_str());
```

"It reports errors that are a consequence of faulty logic within the program such as violating logical preconditions or class invariants and may be preventable."

https://en.cppreference.com/w/cpp/error/logic_error

Maybe that doesn't really describe what happens here, maybe system_error is more like it, but this is not important. I see all of these standard exceptions do get caught.

If there is need to propagate an error beyond this module, better that a subclass of AudacityException be used for that.